pkg/util/purge/purge.go

```diff
@@ -0,0 +1,323 @@
+	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
+	subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID")
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	config := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+	config.Resource = azure.PublicCloud.GraphEndpoint
+
+	rbAuthorizer, err := config.Authorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResourceCleaner{
+		log:    log,
+		dryRun: dryRun,
+
+		resourceRegions: []string{"v4-eastus", "v4-westeurope", "v4-australiasoutheast"},
+
+		resourcegroupscli:      features.NewResourceGroupsClient(subscriptionID, authorizer),
+		vnetscli:               network.NewVirtualNetworksClient(subscriptionID, authorizer),
+		privatelinkservicescli: network.NewPrivateLinkServicesClient(subscriptionID, authorizer),
+		roleassignmentcli:      authorization.NewRoleAssignmentsClient(subscriptionID, authorizer),
+		applicationscli:        graphrbac.NewApplicationsClient(tenantID, rbAuthorizer),
+		serviceprincipalcli:    graphrbac.NewServicePrincipalsClient(tenantID, rbAuthorizer),
+
+		// DeleteCheck decides whether the resource group gets deleted
+		deleteCheck: deleteCheck,
+	}, nil
+}
+
+// CleanResourceGroups loop through the resourgroups in the subscription
+// and deleted everything that is not marked for deletion
+// The deletion check is performed by passed function: DeleteCheck
+func (rc *ResourceCleaner) CleanResourceGroups(ctx context.Context) error {
+	// every resource have to live in the group, therefore deletion clean the unused groups at first
+	gs, err := rc.resourcegroupscli.List(ctx, "", nil)
+	if err != nil {
+		return err
+	}
+
+	sort.Slice(gs, func(i, j int) bool { return *gs[i].Name < *gs[j].Name })
+	for _, g := range gs {
+		err := rc.cleanResourceGroup(ctx, g)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+// cleanResourceGroup checkes whether the resource group can be deleted if yes proceed to clean the group in an order:
+//     - unassign subnets
+//     - clean private links
+//     - checks ARO presence -> store app object ID for futher use
+//     - deletes resource group
+func (rc *ResourceCleaner) cleanResourceGroup(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
+	if rc.deleteCheck(resourceGroup, rc.log) {
+		rc.log.Printf("Deleting ResourceGroup: %s", *resourceGroup.Name)
+		err := rc.cleanNetworking(ctx, resourceGroup)
+		if err != nil {
+			return err
+		}
+
+		err = rc.cleanPrivateLink(ctx, resourceGroup)
+		if err != nil {
+			return err
+		}
+
+		if !rc.dryRun {
+			_, err := rc.resourcegroupscli.Delete(ctx, *resourceGroup.Name)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// cleanNetworking lists subnets in vnets and unnassign security groups
+func (rc *ResourceCleaner) cleanNetworking(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
```

Comment:
I'm not sure this makes sense to me.  Isn't the problematic case where the NSG is in the RG that you want to delete, but it's attached to a vnet in some other RG?  I don't think this function is going to deal with that?